### PR TITLE
ipq40xx: add support for EnGenius EAP1300

### PR DESCRIPTION
**Pull request by gamism**
SOC:    IPQ4018 / QCA Dakota
CPU:    Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
DRAM:   256 MiB
NOR:    32 MiB
ETH:    Qualcomm Atheros QCA8072
WLAN1:  Qualcomm Atheros QCA4018 2.4GHz 802.11bgn 2:2x2
WLAN2:  Qualcomm Atheros QCA4018 5GHz 802.11a/n/ac 2:2x2
INPUT:  RESET Button
LEDS:   Power, LAN, MESH, WLAN 2.4GHz, WLAN 5GHz

1. Load Ramdisk via U-Boot

To set up the flash memory environment, do the following:
a. As a preliminary step, ensure that the board console port is connected to the PC using these RS232 parameters:
   * 115200bps
   * 8N1
b. Confirm that the PC is connected to the board using one of the Ethernet ports. Set a static ip 192.168.99.8 for Ethernet that connects to board. The PC must have a TFTP server launched and listening on the interface to which the board is connected. At this stage power up the board and, after a few seconds, press 4 and then any key during the countdown.

U-BOOT> set fdt_high 0x87000000 & set serverip 192.168.99.8 && set ipaddr 192.168.99.9 && tftpboot 0x84000000 openwrt-ipq40xx-engenius_eap1300-initramfs-fit-uImage.itb && bootm 0x84000000

2. Load image via GUI

a. Upgrade EAP1300 to FW v3.5.3.2
In the GUI, System Manager > Firmware > Firmware Upgrade, to do upgrade.
b. Transfer to OpenWrt from EnGenius.
In Firmware Upgrade page, to upgrade yours openwrt-ipq40xx-engenius_eap1300-squashfs-sysupgrade.bin.

3. Revert to EnGenius EAP1300
To flash openwrt-ipq40xx-engenius_eap1300-squashfs-factory.bin by using sysupgrade command and "DO NOT" keep configuration.
$ sysupgrade –n openwrt-ipq40xx-engenius_eap1300-squashfs-factory.bin

Signed-off-by: Steven Lin <steven.lin@senao.com>

### CODE REVIEW

**Comment by robimarko** [comment]
You specified a ath10k calibration variant but ipq-wifi is not used.
Are BDFs already upstream?

**Comment by gamism** [comment]
> You specified a ath10k calibration variant but ipq-wifi is not used.
> Are BDFs already upstream?

Did you mean this info, "qcom,ath10k-calibration-variant = "EnGenius-EAP1300";" setting in DTS?
Do we have to register an ID for EnGenius EAP1300 model?

**Comment by robimarko** [comment]
Yes, I mean that.
You have to extract BDFs for that device and make a custom board-2.bin with that variant, send it upstream and temporarely use ipq-wifi package in OpenWrt.
Please consult with the merged ipq40xx PRs

**Comment by stefan-kolev** [comment]
Hi,
Sorry for spamming this PR, I read other RP for ipq40xx but I can't understand how they generate the two binary files need by "qca-swiss-army-knife" to make board-2.bin?
There is an example here [https://github.com/openwrt/openwrt/pull/713#issuecomment-369844218](url) 
Thanks!

**Comment by robimarko** [comment]
Simple, you extract the BDFs either by unpacking a stock firmware or from a running system.
There is usually a script that will select which BDF-s to use based on the region, for OpenWrt we usually extract the ETSI or FCC ones as they cover most of the world with their restrictions.

BDF-s are usually in /lib/firmware or /etc folders.

**Comment by gamism** [comment]
> Yes, I mean that.
> You have to extract BDFs for that device and make a custom board-2.bin with that variant, send it upstream and temporarely use ipq-wifi package in OpenWrt.
> Please consult with the merged ipq40xx PRs

I have been updated my pull request that include the ipq-wifi package.
Please help to check if any missing.
Thank you.

Note: I have been send the request to ath10k for update BDF too.
http://lists.infradead.org/pipermail/ath10k/2018-October/012357.html

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 1]
```diff
@@ -151,6 +151,9 @@ case "$FIRMWARE" in
 		ath10kcal_extract "ART" 4096 12064
 		ath10kcal_patch_mac_crc $(macaddr_add $(cat /sys/class/net/eth0/address) -2)
 		;;
+	engenius,eap1300)
```
Comment (robimarko) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: There are other devices that specify the same partition so add it there

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 2]
```diff
@@ -151,6 +151,9 @@ case "$FIRMWARE" in
...
+		ath10kcal_extract "0:ART" 4096 12064
```
Comment (robimarko) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: Like above

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 3]
```diff
@@ -187,6 +190,9 @@ case "$FIRMWARE" in
 		ath10kcal_extract "ART" 20480 12064
 		ath10kcal_patch_mac_crc $(macaddr_add $(cat /sys/class/net/eth0/address) -1)
 		;;
+	engenius,eap1300)
```
Comment (robimarko) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: Again, same as above

File: target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts [hunk 4]
```diff
@@ -0,0 +1,266 @@
+/* Copyright (c) 2015, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2017, Sven Eckelmann <sven.eckelmann@openmesh.com>
```
Comment (robimarko) [inline target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts]: Clearly not you so remove it

File: target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+		mesh {
+			label = "mesh_led";
+			gpios = <&tlmm 1 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+
+		wlan-2g {
+			label = "wlan_2g_led";
+			gpios = <&tlmm 3 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
```
Comment (robimarko) [inline target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts]: Any reason for these two to be on by default?
They are simply WLAN LEDs

File: target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+			};
+			partition@1f50000 {
+				label = "u-boot-env";
+				reg = <0x01f50000 0x00010000>;
+			};
+			partition@1f60000 {
+				label = "userconfig";
+				reg = <0x01f60000 0x000a0000>;
+			};
+			firmware@190000 {
```
Comment (robimarko) [inline target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts]: use partition@ and add label

**Comment by pepe2k** [comment]
@gamism please get rid of Dropbox links inside commit description. They might disappear in future and thus confuse users wanting to upgrade (or revert to vendor firmware) their boards.

Is the firmware `v3.5.3.2` officially and publicly available on vendor (EnGenius) website or is that only a dedicated firmware for OpenWrt migration?

Cheers,
Piotr

**Comment by gamism** [comment]
> @gamism please get rid of Dropbox links inside commit description. They might disappear in future and thus confuse users wanting to upgrade (or revert to vendor firmware) their boards.
> 
> Is the firmware `v3.5.3.2` officially and publicly available on vendor (EnGenius) website or is that only a dedicated firmware for OpenWrt migration?
> 
> Cheers,
> Piotr

Hi @pepe2k ,

I have removed the Dropbox link from commit.  The factory.bin will be created on the OpenWrt wiki to maintain and update.

The firmware `v3.5.3.2` is our internal firmware for OpenWrt migration and not release yet, we have integrated codes in formal firmware, and will officially release on our website.

**Comment by pepe2k** [comment]
@gamism

> The firmware `v3.5.3.2` is our internal firmware for OpenWrt migration and not release yet, we have integrated codes in formal firmware, and will officially release on our website.

I'm suggesting then to wait with merging this PR until the required firmware is officially and publicly available on the vendor website.

**Comment by gamism** [comment]
Hi @robimarko ,

Thank you for reviewing the merge request.
All mentioned changes have been updated.

**Review by chunkeey — COMMENTED** [review]
In the commit message:

> U-BOOT> set fdt_high 0x87000000 & set serverip 192.168.99.8 && 

The single "&" looks a bit out of place. Shouldn't it be a "&&"? 
But is it necessary to set the fdt_high in the first place? Maybe you can remove it
Because OpenWrt uses a different reserved memory map setup than the QSDK.
So the fdt location shouldn't be a problem.

Also, it should be possible to "trim down" the initramfs u-boot command line a bit.
I think this should work, as long as the openwrt.itb is either replaced with a the
long file name or the file is renamed on the tftpserver.

U-BOOT> set ipaddr 192.168.99.8 && tftpboot 84000000 192.168.99.8:openwrt.itb && bootm

Thanks.

**Comment by gamism** [comment]
Hi @chunkeey ,

Thanks for reviewing code and point out the errors.
All mentioned things should be modified and update.

**Comment by gamism** [comment]
Merge conflict solved.
If there are any other errors I should be modify?
Thanks for any comments.

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 7]
```diff
@@ -141,7 +141,9 @@ case "$FIRMWARE" in
 		ath10kcal_extract "ART" 4096 12064
 		ath10kcal_patch_mac_crc $(mtd_get_mac_binary dnidata 0)
 		;;
+	engenius,eap1300 |\
 	compex,wpj428 |\
+	engenius,eap1300 |\
```
Comment (Escalion) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: This seems to be double-defined?

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 8]
```diff
@@ -177,7 +179,9 @@ case "$FIRMWARE" in
 		ath10kcal_extract "ART" 20480 12064
 		ath10kcal_patch_mac_crc $(mtd_get_mac_binary dnidata 12)
 		;;
+	engenius,eap1300 |\
```
Comment (Escalion) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: This entry needs to be removed now that it is between the compex,wpj428 and the openmesh,a42.

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata (hunk 7 above)
Comment (chunkeey) [reply target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: Yes, It looks that way. There are also two patches with the same name
"ipq40xx: add support for EnGenius EAP1300" in the PR as well as a
glue patch "Merge branch 'master' into add-support-for-EnGenius-EAP1300".
I think the OpenWrt project owners are waiting for @gamism to fix the two engenius,eap1300 entries and squash the three patches together into a single "ipq40xx: add support for EnGenius EAP1300" patch.

**Comment by gamism** [comment]
Hi @Escalion ,
The errors have been solved, thank you.

Hi @chunkeey ,
Thanks for your kind reminder.

**Review by chunkeey — APPROVED** [review]
The https://openwrt.org/submitting-patches guideline calls for:
"must have less than 75 characters per line" for the commit message.

However, I don't mind this too much because the one-liner u-boot
scripts and the proper filenames make this impossible (and the github
editor/interface isn't helping either).

Otherwise:

Reviewed-by: Christian Lamparter <chunkeey@gmail.com>

File: package/firmware/ipq-wifi/Makefile [hunk 9]
```diff
@@ -17,7 +17,7 @@ endef
 # Please send a mail with your device-specific board files upstream.
 # You can find instructions and examples on the linux-wireless wiki:
 # <https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles>
-ALLWIFIBOARDS:=
+ALLWIFIBOARDS:=engenius_eap1300
```
[the remark this answers is not in the document]
Comment (gamism) [reply package/firmware/ipq-wifi/Makefile]: Hi Chunkeey,

I have been send the mail to ath10k.

https://www.mail-archive.com/ath10k@lists.infradead.org/msg09444.html

Thanks for the reminder.

**Comment by mkresin** [comment]
merged, thanks!

**Pull request closed without merging**